### PR TITLE
Add JDBC fetch size option for Redshift driver

### LINKED ISSUE #11878: Ablity to set jdbc fetch size for database
**Is your feature request related to a problem? Please describe.**

Most database has smaller default fetch size but not Redshift.

From Redshift documentation- “By default, the JDBC driver collects all the results for a query at one time. As a result, when you attempt to retrieve a large result set over a JDBC connection, you might encounter a client-side out-of-memory error. To enable your client to retrieve result sets in batches instead of in a single all-or-nothing fetch, set the JDBC fetch size parameter in your client application.”

https://docs.aws.amazon.com/redshift/latest/dg/queries-troubleshooting.html#set-the-JDBC-fetch-size-parameter

**Describe the solution you'd like**
We should be able to set a fetch size parameter for a database in metabase. Blank should mean default server settting.

**Describe alternatives you've considered**
Increasing heap size for metabase jvm

**How important is this feature to you?**
This has caused out of memory in metabase jvm. Increasing heap size is possibility and it works up to an extent. 

**Additional context**
Discussion about this in metabase fourm: 
https://discourse.metabase.com/t/is-there-a-way-we-can-set-fetch-size-for-jdbc-for-redshift-database-in-metabase/8880

### PR DESCRIPTION
**Pull request by jeff303**
Originally, this change added a generic option to set the fetch size for a variety of JDBC drivers, on the admin page.  After a lot of discussion (see below), it has been reworked significiantly.

Now, this change aims to (narrowly) solve [this problem](https://docs.aws.amazon.com/redshift/latest/dg/queries-troubleshooting.html#set-the-JDBC-fetch-size-parameter) with the Redshift driver, namely that the fetch size is left to "infinite", which can cause problems with large result sets.  It includes the following:

- A new setting (not exposed in the UI, only controllable via environment variable) called `redshift-fetch-size` that allows people to set the `fetchSize` that will be used for all Redshift queries
- Documentation on this new setting in the environment variables markdown file
- An update of the Redshift JDBC driver dependency to the latest version (2.0.0.3), which is necessary because [the JDBC URL parameter](https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-configuration-options.html#jdbc20-defaultrowfetchsize-option) we now use to control the fetchSize was only introduced in the 2.x version of the driver
- Updates to the Redshift Metabase driver itself to bring in the new version, and to set the URL parameter based on the setting
- Test to ensure that the new URL parameter is set, and also that it actually takes effect (i.e. that the underlying `PreparedStatement` does, in fact, have that value when `.getFetchSize` is called on it)

### CODE REVIEW

**Comment by jeff303** [comment]
Adding @robdaemon for a WIP review (to have a look at the initial direction, before refinement).

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (robdaemon) [inline src/metabase/driver/sql_jdbc/execute.clj:252]: this might be a place to use optional args, like:

`[driver conn sql params canceled-chan & {:keys [fetch-size]}]`

and then when you call it:

`(prepared-statement* driver conn sql params (context/canceled-chan context) :fetch-size fs)`

**Comment by jeff303** [comment]
This is for #11878.

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (jeff303) [reply src/metabase/driver/sql_jdbc/execute.clj]: That is better (fixed).

**Comment by camsaul** [comment]
We should be careful not to add too many random little knobs for random settings. (Ping @salsakran, who has strong opinions on this.) It gets overwhelming from a UX perspective to have to choose values for 50 different settings when connecting your data warehouse during setup. Metabase's differentiator from the competition is ease of use.

Questions to ask yourself before adding a new knob for a database connection option:

- Is this something that can be easily passed in as part of the "additional JDBC connection string parameters" field? A quick search turned up this: https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-configuration-options.html#jdbc20-defaultrowfetchsize-option. So it sounds like the answer would be yes
- Is this something where we should be picking sane default values for users? According to the [Zen of Metabase](https://github.com/metabase/metabase/blob/2a0683bc3dcf1b3e11e41d073c84bac71b3ddce5/zen.md):

   > Don’t ask the user for information the system should already know

   It seems like letting the Redshift JDBC driver fetch a million rows at once is the wrong behavior for anybody, and we shouldn't ask people to set a connection parameter to rectify that. We should pick a sane default value by default and do that for everybody

**Review by camsaul — CHANGES_REQUESTED** [review]
I don't think we need to add a new knob here.

**Comment by jeff303** [comment]
> We should be careful not to add too many random little knobs for random settings. (Ping @salsakran, who has strong opinions on this.) It gets overwhelming from a UX perspective to have to choose values for 50 different settings when connecting your data warehouse during setup. Metabase's differentiator from the competition is ease of use.
> 
> Questions to ask yourself before adding a new knob for a database connection option:
> 
> * Is this something that can be easily passed in as part of the "additional JDBC connection string parameters" field? A quick search turned up this: https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-configuration-options.html#jdbc20-defaultrowfetchsize-option. So it sounds like the answer would be yes

Yeah, in the case of Redshift, it does appear that they allow specifying it in the JDBC URL.  And their [default behavior](https://docs.aws.amazon.com/redshift/latest/dg/queries-troubleshooting.html#set-the-JDBC-fetch-size-parameter) is to fetch **all** rows, which clearly seems to be problematic for us, hence #11878.

> * Is this something where we should be picking sane default values for users? According to the [Zen of Metabase](https://github.com/metabase/metabase/blob/2a0683bc3dcf1b3e11e41d073c84bac71b3ddce5/zen.md):
>   > Don’t ask the user for information the system should already know 

As a general matter, I don't think that is the case that JDBC drivers let you control this through a JDBC URL parameter (Redshift being the fortunate exception).  But some of them have their own unique quirky recommendations (ex: MySQL, which [wants](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-implementation-notes.html) `Integer.MIN_VALUE` if doing an "infinite" streaming type of query; we would need to see what the MariaDB driver does in that regard).  But I take your point that we should just be setting it automatically, so I can back the scope of this completely off and just set the Redshift value to something like the page size (which I think we have, somewhere).  And for any other drivers, we can just tackle them piecemeal as appropriate.

**Comment by camsaul** [comment]
That link I posted (https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-configuration-options.html#jdbc20-defaultrowfetchsize-option) is for the Redshift JDBC documentation, for the `defaultRowFetchSize` option... is that not the same thing?

**Comment by camsaul** [comment]
👍  I think some sort of "page size" or batch size makes sense as a default for everybody. We want our QP to be streaming, although we don't want to do roundtrips to the DB to fetch every row, so it depends on what's happening under the hood. If it's opening a new connection for each batch, it makes sense to use a larger batch size. If it's actually streaming rows, we can use a smaller size like with MySQL for example

**Comment by jeff303** [comment]
Moving back to draft, for now.  Attempting to use the `defaultRowFetchSize` URL parameter led down a rabbit hole:

- This parameter is only supported in the 2.0 version of the Redshift JDBC driver, as explained in the links above.
- We can't use the `2.0.0.2` version of the driver directly, due to [a bug that AWS coincidentally recently fixed](https://github.com/aws/amazon-redshift-jdbc-driver/commit/3c337cb9bb6590fe23a1aa9f5f121a7e29a3655e)
- Once that fix is merged in, and a new release of the Redshift 2.x driver is made, I can update the first commit in this new review to use it and confirm it plays nicely with our uberjar

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (camsaul) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:211]: If it's defined with `defsetting` you have to use `mt/with-temporary-setting-values` to override the value, because it caches stuff internally

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (camsaul) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:210]: Generally it's better to avoid magic numbers like this from a code readability standpoint. It's also one less line that has to change if we change the default value.

You can do something like

```clj
(:default (setting/resolve-setting :redshift-fetch-size))
```

to get the default value

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (camsaul) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:202]: you're leaking a `Connection` here

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (camsaul) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:219]: you don't need to do all this yourself, just use `mt/with-everything-store` in your tests

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (camsaul) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:224]: Instead of defining a new function here just for this one test situation, why not just with-redefs `sql-jdbc.execute/execute-query!` or `execute-reducible-query`? See for example `metabase.query-processor.middleware.cache-test/export-test`. I'd rather not make code more complicated in a general namespace like `sql-jdbc.execute` for the sake of one test in one driver somewhere. It's not a scalable way to do things

**Review by camsaul — CHANGES_REQUESTED** [review]
Overall this looks ok to me but there are a few things to fix like leaking a Connection in the tests and a few other small change suggestions. Ping me when it's ready for another pass, should be ready to merge after the last few tiny tweaks

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (jeff303) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj]: Done.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (jeff303) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj]: Done.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (jeff303) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj]: Oops.  Fixed.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (jeff303) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj]: Changed.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj
Comment (jeff303) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj]: OK, reworked the test to redef `reducible-rows`, which allows us to capture what we need there.

**Review by camsaul — APPROVED** [review]
Cool. This LGTM. We might need to tweak things a little when we get around to implementing #9695, but we can cross that bridge when we get there

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (jeff303) [reply src/metabase/driver/sql_jdbc/execute.clj]: (actually, irrelevant now with latest changes)

**Pull request merged**
